dbms/src/Interpreters/ProbeProcessInfo.h

```diff
@@ -0,0 +1,134 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Columns/ColumnNullable.h>
+#include <Core/Block.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+namespace DB
+{
+enum class CrossProbeMode
+{
+    /// DEEP_COPY_RIGHT_BLOCK construct the probe block using CrossJoinAdder, the right rows will be added to probe block row by row, used when right side has few rows
+    DEEP_COPY_RIGHT_BLOCK,
+    /// SHALLOW_COPY_RIGHT_BLOCK construct the probe block without copy right block, the left row is appended to the right block, used when right side has many rows
+    /// Note for the rows that is filtered by left condition, it still use CrossJoinAdder to construct the probe block
+    SHALLOW_COPY_RIGHT_BLOCK,
+};
+
+struct ProbeProcessInfo
+{
+    Block block;
+    size_t partition_index;
+    UInt64 max_block_size;
+    UInt64 min_result_block_size;
+    size_t start_row;
+    size_t end_row;
+    bool all_rows_joined_finish;
+
+    /// these should be inited before probe each block
+    bool prepare_for_probe_done = false;
+    ColumnPtr null_map_holder = nullptr;
+    ConstNullMapPtr null_map = nullptr;
+    /// Used with ANY INNER ANTI JOIN
+    std::unique_ptr<IColumn::Filter> filter = nullptr;
+    /// Used with ALL ... JOIN
+    std::unique_ptr<IColumn::Offsets> offsets_to_replicate = nullptr;
+
+    /// for hash probe
+    Columns materialized_columns;
+    ColumnRawPtrs key_columns;
+
+    /// for cross probe
+    Block result_block_schema;
+    std::vector<size_t> right_column_index;
+    size_t right_rows_to_be_added_when_matched = 0;
+    CrossProbeMode cross_probe_mode = CrossProbeMode::DEEP_COPY_RIGHT_BLOCK;
+    /// the following fields are used for NO_COPY_RIGHT_BLOCK probe
+    size_t right_block_size = 0;
+    /// the rows that is filtered by left condition
+    size_t row_num_filtered_by_left_condition = 0;
+    size_t next_right_block_index = 0;
+    /// used for outer/semi/anti/left outer semi/left outer anti join
+    bool has_row_matched = false;
+    /// used for left outer semi/left outer anti join
+    bool has_row_null = false;
+
+    explicit ProbeProcessInfo(UInt64 max_block_size_)
+        : partition_index(0)
+        , max_block_size(max_block_size_)
+        , min_result_block_size((max_block_size + 1) / 2)
+        , start_row(0)
+        , end_row(0)
+        , all_rows_joined_finish(true){};
+
+    void resetBlock(Block && block_, size_t partition_index_ = 0);
+    template <bool cross_join>
+    void updateStartRow()
+    {
+        if constexpr (cross_join)
+        {
+            if (cross_probe_mode == CrossProbeMode::SHALLOW_COPY_RIGHT_BLOCK)
```

Comment:
Looks like it can be used as a constexpr variable.
```
template <bool cross_join, bool is_SHALLOW_COPY>
void updateStartRow()
```

Reply:
I found that `cross_join` can be changed to `is_shallow_cross_probe_mode`, so still only one template argument is enough.